src/linux_native.rs

```diff
@@ -0,0 +1,670 @@
+//! This backend uses libudev to discover devices and then talks to hidraw directly
+
+use std::{
+    cell::{Cell, Ref, RefCell},
+    convert::TryInto,
+    ffi::{CStr, CString, OsStr, OsString},
+    fs::{File, OpenOptions},
+    io::Read,
+    os::{
+        fd::{AsRawFd, OwnedFd},
+        unix::{ffi::OsStringExt, fs::OpenOptionsExt},
+    },
+    path::{Path, PathBuf},
+};
+
+use nix::{
+    errno::Errno,
+    poll::{poll, PollFd, PollFlags},
+    sys::stat::{fstat, major, minor},
+    unistd::{read, write},
+};
+
+use super::{
+    ioctl::{hidraw_ioc_get_feature, hidraw_ioc_grdescsize, hidraw_ioc_set_feature},
+    BusType, DeviceInfo, HidDeviceBackendBase, HidError, HidResult, WcharString,
+};
+
+// From linux/hid.h
+const HID_MAX_DESCRIPTOR_SIZE: usize = 4096;
+
+// Bus values from linux/input.h
+const BUS_USB: u16 = 0x03;
+const BUS_BLUETOOTH: u16 = 0x05;
+const BUS_I2C: u16 = 0x18;
+const BUS_SPI: u16 = 0x1C;
+
+pub struct HidApiBackend;
+
+impl HidApiBackend {
+    pub fn get_hid_device_info_vector() -> HidResult<Vec<DeviceInfo>> {
+        // The C version assumes these can't fail, and they should only fail in case
+        // of memory allocation issues, at which point maybe we should panic
+        let mut enumerator = match udev::Enumerator::new() {
+            Ok(e) => e,
+            Err(_) => return Ok(Vec::new()),
+        };
+        enumerator.match_subsystem("hidraw").unwrap();
+        let scan = match enumerator.scan_devices() {
+            Ok(s) => s,
+            Err(_) => return Ok(Vec::new()),
+        };
+
+        let mut devices = Vec::new();
```

Comment:
We return a vec so we still need a flatten after this. I suppose we could return an empty vector instead of a None and then a `flatten_map` by itself would work, but the semantics are slightly different then.